src/config.h

```diff
@@ -100,9 +100,9 @@
 /* Test for accept4() */
 #if defined(__linux__) || \
     defined(OpenBSD5_7) || \
-    __DragonFly_version >= 400305 || \
-    (__FreeBSD__ >= 10 || __FreeBSD_version >= 1000000) || \
-    (defined(NetBSD8_0) || __NetBSD_Version__ >= 800000000)
+    (defined(__DragonFly__) && __DragonFly_version >= 400305) || \
+    (defined(__FreeBSD__) && (__FreeBSD__ >= 10 || __FreeBSD_version >= 1000000)) || \
```

Comment:
I'm fine with that, but I'm not sure what the minimal version of FreeBSD supported by `valkey`. There is another detection for FreeBSD at the moment:
https://github.com/valkey-io/valkey/blob/52f9291f797a7f24267115bedb396f16310456ed/src/config.h#L318-L321
 If `valkey` has decided to drop the support of FreeBSD <10.0, then we can go for it.

Reply:
@madolson @zuiderkwast Some advices?

Reply:
What's the point of adding `defined()`? An undefined macro evaluates to 0 in `#if`.

I fail to see the point of this PR.

Reply:
Add the missing <param.h> and add detection for dragonflybsd

Reply:
> An undefined macro evaluates to 0 in #if

Yes, but it could emit warnings for that, I think it's nice to eliminate that.

Reply:
Right, OK.

Dragonfly doesn't work without sys/param.h but all other systems work without it? Why is that?

Reply:
The last release prior to FreeBSD 10.0 is 9.3, which passed EOL as of December 31, 2016. It's certainly reasonable to desupport anything under 10.0.

https://www.freebsd.org/security/unsupported/

Reply:
Some systems define those macros in <sys/param.h>: https://sourceforge.net/p/predef/wiki/OperatingSystems/

Reply:
> Dragonfly doesn't work without sys/param.h but all other systems work without it? Why is that?

`__FreeBSD__` is a built-in compiler define, while `__FreeBSD_version` comes from sys/param.h; `__FreeBSD__ >= 10` is true regardless of whether sys/param.h is included or not. Similarly `__DragonFly_Version` comes from sys/param.h.

These are the built-in compiler defines on my system:
```
$ cc -x c -dM -E /dev/null | grep FreeBSD
#define __FreeBSD__ 14
#define __FreeBSD_cc_version 1400006
#define __VERSION__ "FreeBSD Clang 16.0.6 (https://github.com/llvm/llvm-project.git llvmorg-16.0.6-0-g7cbf1a259152)"
```

Reply:
Interesting. So without `sys/param.h` the checks for OpenBSD (`OpenBSD5_7`) and NetBSD (`NetBSD8_0`, `__NetBSD_Version__`) cannot have worked?

Reply:
I don't really care if we remove or keep the check for FreeBSD 10. It's on the same line and it doesn't harm anyone. You can do as you want. :)

Reply:
> Interesting. So without `sys/param.h` the checks for OpenBSD (`OpenBSD5_7`) and NetBSD (`NetBSD8_0`, `__NetBSD_Version__`) cannot have worked?

Yes, check out: [NetBSD param.h](https://github.com/NetBSD/src/blob/a5de3f67cd5b6f6eea88c5e0e82e2805b751d40c/sys/sys/param.h#L70), [OpenBSD param.h](https://github.com/openbsd/src/blob/39e743a585bfb0027efb6a4bb6f3b6b02f8ed1e3/sys/sys/param.h#L44), [DragonFlyBSD param.h](https://github.com/DragonFlyBSD/DragonFlyBSD/blob/7eb3d59f73026ff7c8b6d53cbea198c972753494/sys/sys/param.h#L257)

Reply:
> I don't really care if we remove or keep the check for FreeBSD 10. It's on the same line and it doesn't harm anyone. You can do as you want. :)

In that case, I'm about to go with what @emaste has suggested.

Reply:
Great fix then.

Why do we need to check the version in two ways for each system? Like `__FreeBSD__ >= 10 || __FreeBSD_version >= 1000000`, I would assume checking one of them would be enough? Same with the other similar checks.

Reply:
Note that they're not exactly equivalent, because the compiler built-in `__FreeBSD__` represents the target the compiler was built for, while `__FreeBSD_version` represents the version from the target headers. They should match (and will match when installing some version of FreeBSD and using the compiler that it comes with), but in some cross-compilation or other atypical configurations they could differ. For example, starting FreeBSD 9, and then updating the kernel and userland but not the compiler would have `__FreeBSD__` == 9 and `__FreeBSD_version` == 10xxxxxx.

Reply:
It's great to have an expert in this thread. :smile: